### ISSUE TITLE
failed to ping localhost:25565 : All attempts fail: #1: failed to connect: dial tcp 127.0.0.1:25565: connect: connection refuse

### ISSUE THREAD

**Issue by mindphaseproductions** [itzg/docker-minecraft-server#1758]
### Describe the problem

In the 'Container health' section within the minecraft_server container in my Portainer, I am receiving an error: "failed to ping localhost:25565 : All attempts fail: #1: failed to connect: dial tcp 127.0.0.1:25565: connect: connection refused". This seems to have not been a problem, as I have had this image running for over a month and have not touched anything. I never even configured a Container health check for any of my containers. Maybe it's due to a recent update?

### Container logs

```
services:
  minecraft_server:
    container_name: minecraft_server
    environment:
      EULA: "TRUE"
      GID: "1000"
      TYPE: VANILLA
      TZ: America/Toronto
      UID: "1000"
      VERSION: LATEST
    hostname: XXXXX
    image: itzg/minecraft-server
    logging:
      driver: json-file
      options:
        max-file: "10"
        max-size: 200k
    networks:
      default: null
    ports:
    - mode: ingress
      target: 25565
      published: "25565"
      protocol: tcp
    restart: unless-stopped
    volumes:
    - type: bind
      source: /etc/localtime
      target: /etc/localtime
      read_only: true
      bind:
        create_host_path: true
    - type: bind
      source: /home/XXXXX/.config/appdata/minecraft_server
      target: /data
      bind:
        create_host_path: true
    - type: bind
      source: /media/storage/XXXXX/storage
      target: /storage
      bind:
        create_host_path: true
  portainer:
    command:
    - -H
    - unix:///var/run/docker.sock
    container_name: portainer
    environment:
      TZ: America/Toronto
    hostname: XXXXX
    image: portainer/portainer-ce
    logging:
      driver: json-file
      options:
        max-file: "10"
        max-size: 200k
    networks:
      default: null
    ports:
    - mode: ingress
      target: 9000
      published: "9000"
      protocol: tcp
    restart: unless-stopped
    volumes:
    - type: bind
      source: /etc/localtime
      target: /etc/localtime
      read_only: true
      bind:
        create_host_path: true
    - type: bind
      source: /var/run/docker.sock
      target: /var/run/docker.sock
      bind:
        create_host_path: true
    - type: bind
      source: /home/XXXXX/.config/appdata/portainer
      target: /data
      bind:
        create_host_path: true
    - type: bind
      source: /media/storage/XXXXX/storage
      target: /storage
      bind:
        create_host_path: true

```

**Comment by itzg** [comment]
Can you please re-paste the container definition within the three backtick blocks. The indentation is unreadable right now. 

Ingress mode on the port mapping is probably the problem. Perhaps it broke with an upgrade of Docker, Portainer, etc.

**Comment by mindphaseproductions** [comment]
> Can you please re-paste the container definition within the three backtick blocks. The indentation is unreadable right now.
> 
> Ingress mode on the port mapping is probably the problem. Perhaps it broke with an upgrade of Docker, Portainer, etc.

Done. Sorry about that.

**Comment by goloroden** [comment]
I have the very same problem, but without Portainer.

I try to run with type `CURSEFORGE`, but I'm running into the same issue: The container goes to `unhealthy` after a while, and if I look at the details using `docker inspect` it tells me (I'm only showing the relevant snippet):

```
        "Health": {
                "Status": "unhealthy",
                "FailingStreak": 26,
                "Log": [
                    {
                        "Start": "2022-10-14T18:49:51.825405467Z",
                        "End": "2022-10-14T18:49:51.938506Z",
                        "ExitCode": 1,
                        "Output": "failed to ping localhost:25565 : could not connect to Minecraft server: dial tcp 127.0.0.1:25565: connect: connection refused"
                    },
```

I originally ran the container on host port 25567, but I've now switched to 25565, but with  no effect.

What's interesting: I have the exact same setup up and running, just without the `TYPE` set (i.e. a vanilla server), in two instances, and both work fine without any problems.

Do you need anything else from me, @itzg, to be able to dig deeper into this problem?

**Comment by itzg** [comment]
@goloroden can you locate and provide a log line that looks like this near the start of the server itself:

```
[20:38:45] [Server thread/INFO]: Starting Minecraft server on *:25565
```

On a vanilla server it'll be in amongst lines like:
```
[20:38:45] [Server thread/INFO]: Starting minecraft server version 1.19.2
[20:38:45] [Server thread/INFO]: Loading properties
[20:38:45] [Server thread/INFO]: Default game type: SURVIVAL
[20:38:45] [Server thread/INFO]: Generating keypair
[20:38:45] [Server thread/INFO]: Starting Minecraft server on *:25565
[20:38:45] [Server thread/INFO]: Using epoll channel type
```

**Comment by goloroden** [comment]
Hey @itzg 👋 

First of all: Thanks for replying that quickly, I didn't expect that – but it's great! Thanks 😊

I just had a look at the logs of the server, and tbh they are pretty short – so apparently the server does not even get to the point where it is starting. The logs look like this:

```
[init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 10 1000 1000 4096 Oct 14 18:22 /data'
[init] Resolved version given LATEST into 1.19.2 and major version 1.19
[init] Resolving type given CURSEFORGE
[init] Looking for Feed-The-Beast / CurseForge server modpack.
[init] detected FTB, changing properties path to /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2/server.properties
[init] Checking for JSON files.
[init] Setting initial memory to 1G and max to 1G
[init] Running FTB /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2/start.sh in /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2 ...

Running Fabric checks and setup...
Improved Fabric Server Launcher available...
The improved launcher will be used to run this Fabric server.
fabric-server-launcher.jar present.

WARNING! The current location of this script contains spaces. This may cause this server to crash!
It is strongly recommended to move this server pack to a location whose path does NOT contain SPACES!

Current path:
/data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2
```

That's it, unfortunately.

The command I'm using to run the container is:

```shell
docker run -d -p 25565:25565 --name mc-better --restart always -e EULA=TRUE -e TYPE=CURSEFORGE -e CF_SERVER_MOD=BMC-Server-1.19.2-v8.5.zip -v /root/minecraft-better/data:/data itzg/minecraft-server
```

I assume that I did something wrong with the `TYPE` and `CF_SERVER_MOD` environment variables, or that something isn't matching. I downloaded the mod from: https://www.curseforge.com/minecraft/modpacks/better-mc-plus-new

TBH I'm not sure if `CURSEFORGE` is correct as type, since although the download was done from curseforge.com, the extracted file says something about `Feed The Beast`, which is – as I've seen – another type. I also read that you could try to just run it as `FORGE`, and I'm a little bit lost here …

Any ideas on this?

**Comment by itzg** [comment]
@goloroden Your approach is good. CF and FTB used to be one hosted together in same format. Some modpack start scripts don't behave right, so instead you could try the generic pack approach like https://github.com/itzg/docker-minecraft-server/blob/master/examples/docker-compose-generic-pack.yml

**Comment by goloroden** [comment]
@itzg Thanks a lot (again) for your very quick response.

I tried it using the following command line:

```shell
docker run -d -p 25565:25565 --name mc-better --restart always -e EULA=TRUE -e TYPE=FORGE -e FORGE_VERSION=37.0.90 -e GENERIC_PACK=BMC-Server-1.19.2-v8.5.zip -v /root/minecraft-better/data:/data itzg/minecraft-server
```

The result of this is a ever-restarting service in the startup phase – it never gets to the point where it's either healthy or unhealthy.

Any thought on this?

**Comment by lecerealkiller** [comment]
> Hey @itzg 👋
> 
> First of all: Thanks for replying that quickly, I didn't expect that – but it's great! Thanks 😊
> 
> I just had a look at the logs of the server, and tbh they are pretty short – so apparently the server does not even get to the point where it is starting. The logs look like this:
> 
> ```
> [init] Running as uid=1000 gid=1000 with /data as 'drwxrwxr-x 10 1000 1000 4096 Oct 14 18:22 /data'
> [init] Resolved version given LATEST into 1.19.2 and major version 1.19
> [init] Resolving type given CURSEFORGE
> [init] Looking for Feed-The-Beast / CurseForge server modpack.
> [init] detected FTB, changing properties path to /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2/server.properties
> [init] Checking for JSON files.
> [init] Setting initial memory to 1G and max to 1G
> [init] Running FTB /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2/start.sh in /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2 ...
> 
> Running Fabric checks and setup...
> Improved Fabric Server Launcher available...
> The improved launcher will be used to run this Fabric server.
> fabric-server-launcher.jar present.
> 
> WARNING! The current location of this script contains spaces. This may cause this server to crash!
> It is strongly recommended to move this server pack to a location whose path does NOT contain SPACES!
> 
> Current path:
> /data/FeedTheBeast/Better MC Server Pack FABRIC 1.19.2
> ```
> 
> That's it, unfortunately.
> 
> The command I'm using to run the container is:
> 
> ```shell
> docker run -d -p 25565:25565 --name mc-better --restart always -e EULA=TRUE -e TYPE=CURSEFORGE -e CF_SERVER_MOD=BMC-Server-1.19.2-v8.5.zip -v /root/minecraft-better/data:/data itzg/minecraft-server
> ```
> 
> I assume that I did something wrong with the `TYPE` and `CF_SERVER_MOD` environment variables, or that something isn't matching. I downloaded the mod from: https://www.curseforge.com/minecraft/modpacks/better-mc-plus-new
> 
> TBH I'm not sure if `CURSEFORGE` is correct as type, since although the download was done from curseforge.com, the extracted file says something about `Feed The Beast`, which is – as I've seen – another type. I also read that you could try to just run it as `FORGE`, and I'm a little bit lost here …
> 
> Any ideas on this?

@goloroden  I was having the same problem. I went into the Better+MC+Server+Pack+FORGE+1.16.5+v59.5.zip and changed the folder name to be BetterMC_FORGE_1.16.5 (removing spaces) and it extracted to the new folder & then I didn't get that message 'WARNING! The current location of this script contains spaces' and it loaded properly.

Looking at start.sh, there is a prompt 'Are you sure you want to continue? (Yes/No):' after that WARNING!, so that's where it's hanging in docker.
